dbuf/dbuf/dbuf.py

```diff
@@ -0,0 +1,287 @@
+            buf[0:n] = self.buf[pos:(pos+n)]
+            return
+        writer = self.writers.get(file_id)
+        if writer is not None:
+            await writer
+            assert not writer.cancelled()
+            assert writer.exception() is None, writer.exception
+        async with self.file_budget:
+            with await self.aiofiles.open(f'{self.data_dir}/{self.id}/{file_id}', 'rb', buffering=0) as f:
+                await self.aiofiles.seek(f, pos)
+                assert len(buf) == n
+                assert await self.aiofiles.readinto(f, buf) == n
+
+    async def readmany(self, keys):
+        out = bytearray(sum(4 + x[2] for x in keys))
+        offs = [0] * len(keys)
+        s = 0
+        for i, n in enumerate(x[2] for x in keys):
+            offs[i] = s
+            s += n + 4
+
+        async def read_into(key, off):
+            _, _, n = key
+            struct.pack_into('i', out, off, n)
+            off += 4
+            await self.read(key, memoryview(out)[off:(off+n)])
+        await asyncio.gather(*[read_into(key, off) for key, off in zip(keys, offs)])
+        return out
+
+    async def delete(self):
+        for i in range(self.current_file):
+            await self.aiofiles.remove(f'{self.data_dir}/{self.id}/{i}')
+        await self.aiofiles.rmdir(f'{self.data_dir}/{self.id}')
+
+
+class Sessions:
+    @staticmethod
+    async def make(bufsize, aiofiles, data_dir):
+        await aiofiles.mkdir(data_dir)
+        return Sessions(bufsize, data_dir, aiofiles)
+
+    def __init__(self, bufsize, data_dir, aiofiles):
+        # consensus
+        self.next_session = 0
+        # local
+        self.bufsize = bufsize
+        self.data_dir = data_dir
+        self.aiofiles = aiofiles
+        self.sessions = {}
+
+    async def new_session(self):
+        id = self.next_session
+        self.next_session += 1
+        self.sessions[id] = await Session.make(self.bufsize, self.data_dir, id, self.aiofiles)
+        return id
+
+    async def delete_session(self, id):
+        await self.sessions[id].delete()
+        del self.sessions[id]
+
+    async def delete(self):
+        for id in self.sessions:
+            await self.delete_session(id)
+        await self.aiofiles.rmdir(self.data_dir)
+
+
+class Server:
+    def __init__(self, name, binding_host, port, leader, dbuf, aiofiles):
+        self.deploy_config = get_deploy_config()
+        self.app = web.Application(client_max_size=50 * 1024 * 1024)
+        self.routes = web.RouteTableDef()
+        self.workers = set()
+        self.name = name
+        self.binding_host = binding_host
+        self.port = port
+        self.dbuf = dbuf
+        self.leader = leader
+        self.leader_url = self.deploy_config.base_url(leader)
+        self.aiofiles = aiofiles
+        self.shuffle_create_lock = asyncio.Lock()
+        self.app.add_routes([
+            web.post('/s', self.create),
+            web.post('/s/{session}', self.post),
```

Comment:
Calling something post seems a bit too generic.  post_session?

Reply:
I changed the name of `post_worker` to `register_worker`. Changing `post` to `post_session` creates asymmetry with `get`. Changing both to `get_session` seems unnecessarily verbose given the context.

I could change this one to `put` and the other to `get`, but there's a verb called PUT so that seems confusing. I could make this `set` instead of `post`, but that seems to imply that I'm modifying something with identity, which I am not.

I really feel like post and get are the right words. Post some data, get some data.

This is not the endpoint that creates a new session, that's `create` / `POST /s`. This endpoint posts data *into* an extant session, `post` / `POST /s/session`